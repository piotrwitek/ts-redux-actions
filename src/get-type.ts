export interface TypeGetter<T> { getType?: () => T }

export function getType<T extends string>(
  actionCreator: ((...args: any[]) => { type: T }) & TypeGetter<T>,
): T {
  if (actionCreator == null) {
    throw new Error('first argument is missing');
  }

  if (actionCreator.getType == null) {
    throw new Error('first argument is not an instance of "typesafe-actions"');
  }

  return actionCreator.getType();
}
