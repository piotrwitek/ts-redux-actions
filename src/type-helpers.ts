/**
 * PUBLIC API
 */

/**
 * @desc Type representing Type Constant
 */
export type TypeConstant = string;

/**
 * @desc Type representing Generic Action
 */
export type Action<TType extends TypeConstant = TypeConstant> = {
  type: TType;
};

/**
 * @desc Type representing Generic ActionCreator
 */
export type ActionCreator<TType extends TypeConstant> = (
  ...args: any[]
) => Action<TType>;

/**
 * @desc Type representing Generic Reducer
 */
export type Reducer<TState, TAction extends Action> = (
  state: TState | undefined,
  action: TAction
) => TState;

/**
 * @desc Action without Payload
 */
export type EmptyAction<TType extends TypeConstant> = {
  type: TType;
};

/**
 * @desc Action with only Payload
 */
export type PayloadAction<TType extends TypeConstant, TPayload> = {
  type: TType;
  payload: TPayload;
};

/**
 * @desc Action with only Meta
 */
export type MetaAction<TType extends TypeConstant, TMeta> = {
  type: TType;
  meta: TMeta;
};

/**
 * @desc Action with both Payload and Meta
 */
export type PayloadMetaAction<TType extends TypeConstant, TPayload, TMeta> = {
  type: TType;
  payload: TPayload;
  meta: TMeta;
};

/**
 * @desc Action Creator producing EmptyAction
 */
export type EmptyAC<TType extends TypeConstant> = () => EmptyAction<TType>;

/**
 * @desc Action Creator producing PayloadAction
 */
export type PayloadAC<TType extends TypeConstant, TPayload> = (
  payload: TPayload
) => PayloadAction<TType, TPayload>;

/**
 * @desc Action Creator producing PayloadMetaAction
 */
export type PayloadMetaAC<TType extends TypeConstant, TPayload, TMeta> = (
  payload: TPayload,
  meta: TMeta
) => PayloadMetaAction<TType, TPayload, TMeta>;

/**
 * @desc Type representing type getter on Action Creator instance
 */
export interface TypeMeta<TType extends TypeConstant> {
  getType?: () => TType;
}

/**
 * @desc Infers Action union-type from action-creator map object
 */
export type ActionType<
  TActionCreatorOrMap extends any
> = TActionCreatorOrMap extends ActionCreator<TypeConstant>
  ? ReturnType<TActionCreatorOrMap>
  : TActionCreatorOrMap extends Record<any, any>
  ? {
      [K in keyof TActionCreatorOrMap]: ActionType<TActionCreatorOrMap[K]>
    }[keyof TActionCreatorOrMap]
  : TActionCreatorOrMap extends infer R // TODO: should be just never but compiler yell with circularly references itself error
  ? never
  : never;

/**
 * @desc Infers State object from reducer map object
 */
export type StateType<
  TReducerOrMap extends any
> = TReducerOrMap extends Reducer<any, any>
  ? ReturnType<TReducerOrMap>
  : TReducerOrMap extends Record<any, any>
  ? { [K in keyof TReducerOrMap]: StateType<TReducerOrMap[K]> }
  : never;

/**
 * INTERNAL API
 */

/** @private */
export type ActionBuilderConstructor<
  TType extends TypeConstant,
  TPayload extends any = undefined,
  TMeta extends any = undefined
> = [TMeta] extends [undefined]
  ? [TPayload] extends [undefined]
    ? unknown extends TPayload
      ? PayloadAC<TType, TPayload>
      : unknown extends TMeta
      ? PayloadMetaAC<TType, TPayload, TMeta>
      : EmptyAC<TType>
    : PayloadAC<TType, TPayload>
  : PayloadMetaAC<TType, TPayload, TMeta>;

/** @private */
export type ActionBuilderMap<
  TType extends TypeConstant,
  TActionProps extends any,
  TPayloadArg extends any = undefined,
  TMetaArg extends any = undefined
> = [TMetaArg] extends [undefined]
  ? [TPayloadArg] extends [undefined]
    ? () => { type: TType } & TActionProps
    : (payload: TPayloadArg) => { type: TType } & TActionProps
  : (payload: TPayloadArg, meta: TMetaArg) => { type: TType } & TActionProps;
