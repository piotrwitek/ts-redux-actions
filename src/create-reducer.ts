import { getType } from './get-type';
import {
  checkValidActionCreator,
  checkValidActionType,
  throwInvalidActionTypeOrActionCreator,
} from './utils/validation';
import { Reducer, Action, Types } from './type-helpers';

type CreateReducerChainApi<
  TState,
  TPrevNotHandledAction extends Action,
  TRootAction extends Action
> = <
  TType extends TPrevNotHandledAction['type'],
  TCreator extends (...args: any[]) => TPrevNotHandledAction,
  TNextNotHandledAction extends Exclude<
    TPrevNotHandledAction,
    Action<TType> & ReturnType<TCreator>
  >,
  TAction extends TPrevNotHandledAction extends Action<TType>
    ? TPrevNotHandledAction extends ReturnType<TCreator>
      ? TPrevNotHandledAction
      : never
    : never
>(
  singleOrMultipleCreatorsAndTypes: TType | TType[] | TCreator | TCreator[],
  reducer: (state: TState, action: TAction) => TState
) => [TNextNotHandledAction] extends [never]
  ? Reducer<TState, TRootAction> & {
      handlers: Record<
        TRootAction['type'],
        (state: TState, action: TRootAction) => TState
      >;
    }
  : Reducer<TState, TRootAction> & {
      handlers: Record<
        Exclude<TRootAction, TNextNotHandledAction>['type'],
        (state: TState, action: TRootAction) => TState
      >;
      handleAction: CreateReducerChainApi<
        TState,
        TNextNotHandledAction,
        TRootAction
      >;
    };

type GetAction<
  TAction extends Action,
  TType extends TAction['type']
> = TAction extends Action<TType> ? TAction : never;

type InitialHandler<TState, TRootAction extends Action> = {
  [P in TRootAction['type']]?: (
    state: TState,
    action: GetAction<TRootAction, P>
  ) => TState;
};

type RootAction = Types extends { RootAction: infer T } ? T : any;

export function createReducer<TState, TRootAction extends Action = RootAction>(
  initialState: TState,
  initialHandlers: InitialHandler<TState, TRootAction> = {}
) {
  const handlers: any = {
    ...initialHandlers,
  };

  const rootReducer: Reducer<TState, TRootAction> = (
    state = initialState,
    action: TRootAction
  ) => {
    if (handlers.hasOwnProperty(action.type)) {
      const reducer = handlers[action.type];
      if (typeof reducer !== 'function') {
        throw Error(
          `Reducer under "${action.type}" key is not a valid reducer`
        );
      }
      return reducer(state, action);
    } else {
      return state;
    }
  };

  const handleAction = ((singleOrMultipleCreatorsAndTypes, reducer) => {
    const creatorsAndTypes = Array.isArray(singleOrMultipleCreatorsAndTypes)
      ? singleOrMultipleCreatorsAndTypes
      : [singleOrMultipleCreatorsAndTypes];

    const newHandlers: typeof handlers = {};
    creatorsAndTypes
      .map(acOrType =>
        checkValidActionCreator(acOrType)
          ? getType(acOrType)
          : checkValidActionType(acOrType)
          ? acOrType
          : throwInvalidActionTypeOrActionCreator()
      )
      .forEach(type => (newHandlers[type] = reducer));

    return createReducer<TState, TRootAction>(initialState, {
      ...handlers,
      ...newHandlers,
    });
  }) as CreateReducerChainApi<TState, TRootAction, TRootAction>;

  const chainApi = Object.assign(rootReducer, {
    handlers: { ...handlers },
    handleAction,
  } as const);

  return chainApi;
}
