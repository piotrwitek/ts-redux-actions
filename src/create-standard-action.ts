import { StringType, Box, FsaBuilder, FsaMapBuilder } from './type-helpers';
import { createCustomAction } from './create-custom-action';
import {
  checkIsEmpty,
  throwIsEmpty,
  checkInvalidActionType,
  throwInvalidActionType,
} from './utils/validation';

export interface CreateStandardAction<T extends StringType> {
  <P = void, M = void>(): FsaBuilder<T, Box<P>, Box<M>>;
  map<R, P = void, M = void>(
    fn: (payload: P, meta: M) => R
  ): FsaMapBuilder<T, Box<R>, Box<P>, Box<M>>;
}

/**
 * @description create an action-creator of a given function that contains hidden "type" metadata
 */
export function createStandardAction<T extends StringType>(
  type: T
): CreateStandardAction<T> {
  if (checkIsEmpty(type)) {
    throwIsEmpty(1);
  }

  if (checkInvalidActionType(type)) {
    throwInvalidActionType(1);
  }

  function constructor<P, M = void>(): FsaBuilder<T, Box<P>, Box<M>> {
    return createCustomAction(type, _type => (payload: P, meta: M) => ({
      type: _type,
      payload,
      meta,
    })) as FsaBuilder<T, Box<P>, Box<M>>;
  }

  function map<R, P, M>(
    fn: (payload: P, meta: M) => R
  ): FsaMapBuilder<T, Box<R>, Box<P>, Box<M>> {
    return createCustomAction(type, _type => (payload: P, meta: M) =>
      Object.assign(fn(payload, meta), { type: _type })
    ) as FsaMapBuilder<T, Box<R>, Box<P>, Box<M>>;
  }

  return Object.assign(constructor, { map });
}
