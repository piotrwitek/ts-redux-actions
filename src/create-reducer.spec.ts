import * as T from './type-helpers'; // type-tests global
import { ActionType } from './type-helpers';
import { createStandardAction } from './create-standard-action';
import { createReducer } from './create-reducer';
import { getType } from './get-type';

const add = createStandardAction('ADD')<number>();
const increment = createStandardAction('INCREMENT')();
const decrement = createStandardAction('DECREMENT')();
const actions = {
  add,
  increment,
  decrement,
};

declare module './type-helpers' {
  interface Types {
    RootAction: ActionType<typeof actions>;
  }
}

const initialState = 0;

// @dts-jest:group With Action Creators
{
  const emptyReducer = createReducer(initialState);

  const counterReducer1 = emptyReducer.handleAction(
    [add, increment],
    (state, action) => state + (action.type === 'ADD' ? action.payload : 1)
  );
  // @dts-jest:pass:snap
  counterReducer1.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer1.handlers }); // => ["ADD", "INCREMENT"]
  // @dts-jest:pass
  Object.keys({ ...emptyReducer.handlers }); // => []

  const counterReducer2 = emptyReducer
    .handleAction([add], (state, action) => state + action.payload)
    .handleAction([increment], (state, _) => state + 1);
  // @dts-jest:pass:snap
  counterReducer2.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer2.handlers }); // => ["ADD", "INCREMENT"]
  // @dts-jest:pass
  Object.keys({ ...emptyReducer.handlers }); // => []

  const counterReducer3 = emptyReducer.handleAction(
    add,
    (state, action) => state + action.payload
  );
  // @dts-jest:pass:snap
  counterReducer3.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer3.handlers }); // => [ "ADD"]
  // @dts-jest:pass
  Object.keys({ ...emptyReducer.handlers }); // => []

  const counterReducer4 = emptyReducer.handleAction(
    increment,
    (state, _) => state + 1
  );
  // @dts-jest:pass:snap
  counterReducer4.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer4.handlers }); // => [ "INCREMENT"]
  // @dts-jest:pass
  Object.keys({ ...emptyReducer.handlers }); // => []

  {
    [
      counterReducer1,
      counterReducer2,
      createReducer(initialState, {
        [getType(add)]: counterReducer3.handlers.ADD,
        [getType(increment)]: counterReducer4.handlers.INCREMENT,
      }),
      createReducer(initialState, {
        [getType(add)]: (state: any, action: any) => state + action.payload,
        [getType(increment)]: (state: any, _: any) => state + 1,
      }),
      createReducer<number, ActionType<typeof actions>>(initialState, {
        [getType(add)]: (state: any, action: any) => state + action.payload,
        [getType(increment)]: (state: any, _: any) => state + 1,
      }),
    ].forEach(fn => {
      // @dts-jest:pass
      fn(0, {} as any); // => 0
      // @dts-jest:pass
      fn(0, increment()); // => 1
      // @dts-jest:pass
      fn(0, add(4)); // => 4
    });
  }
}

// @dts-jest:group Type refinement checks
{
  type State = {
    foo: string | null;
  };

  const defaultState: State = {
    foo: null,
  };

  const actions2 = {
    foo1: createStandardAction('foo1')<string>(),
    foo2: createStandardAction('foo2')<string>(),
    foo3: createStandardAction('foo3')(),
    foo4: createStandardAction('foo4')(),
  };

  type Action = ActionType<typeof actions2>;

  const reducer = createReducer<State, Action>(defaultState)
    .handleAction(actions2.foo1, (state, action) => {
      // @dts-jest:pass:snap
      state;
      // @dts-jest:pass:snap
      action;

      return {
        ...state,
        foo: action.payload,
      };
    })
    .handleAction(actions2.foo2, (state, action) => {
      // @dts-jest:pass:snap
      state;
      // @dts-jest:pass:snap
      action;

      return {
        ...state,
        foo: action.payload,
      };
    })
    .handleAction(actions2.foo3, (state, action) => {
      // @dts-jest:pass:snap
      state;
      // @dts-jest:pass:snap
      action;

      return {
        ...state,
        foo: 'empty',
      };
    })
    .handleAction(actions2.foo4, (state, action) => {
      // @dts-jest:pass:snap
      state;
      // @dts-jest:pass:snap
      action;

      return {
        ...state,
        foo: 'empty',
      };
    });

  // @dts-jest:pass:snap
  reducer.handlers;

  {
    [
      reducer(defaultState, actions2.foo1('check')),
      reducer(defaultState, actions2.foo2('check')),
    ].forEach(reducerResult => {
      // @dts-jest:pass:snap
      reducerResult; // => { foo: "check" }
    });

    [
      reducer(defaultState, actions2.foo3()),
      reducer(defaultState, actions2.foo4()),
    ].forEach(reducerResult => {
      // @dts-jest:pass:snap
      reducerResult; // => { foo: "empty" }
    });
  }
}

// @dts-jest:group With Action Types
{
  const reducerTest = createReducer(initialState);

  const counterReducer1 = reducerTest.handleAction(
    ['ADD', 'INCREMENT'],
    (state, action) => state + (action.type === 'ADD' ? action.payload : 1)
  );
  // @dts-jest:pass:snap
  counterReducer1.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer1.handlers }); // => ["ADD", "INCREMENT"]
  // @dts-jest:pass
  Object.keys({ ...reducerTest.handlers }); // => []

  const counterReducer2 = reducerTest
    .handleAction(['ADD'], (state, action) => state + action.payload)
    .handleAction(['INCREMENT'], (state, _) => state + 1);
  // @dts-jest:pass:snap
  counterReducer2.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer2.handlers }); // => ["ADD", "INCREMENT"]
  // @dts-jest:pass
  Object.keys({ ...reducerTest.handlers }); // => []

  const counterReducer3 = reducerTest.handleAction(
    'ADD',
    (state, action) => state + action.payload
  );
  // @dts-jest:pass:snap
  counterReducer3.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer3.handlers }); // => ["ADD"]
  // @dts-jest:pass
  Object.keys({ ...reducerTest.handlers }); // => []

  const counterReducer4 = reducerTest.handleAction(
    'INCREMENT',
    (state, _) => state + 1
  );
  // @dts-jest:pass:snap
  counterReducer4.handlers;
  // @dts-jest:pass
  Object.keys({ ...counterReducer4.handlers }); // => [ "INCREMENT"]
  // @dts-jest:pass
  Object.keys({ ...reducerTest.handlers }); // => []

  {
    [
      counterReducer1,
      counterReducer2,
      createReducer(initialState, {
        ADD: counterReducer3.handlers.ADD,
        INCREMENT: counterReducer4.handlers.INCREMENT,
      }),
      createReducer(initialState, {
        ADD: (state, action) => state + action.payload,
        INCREMENT: (state, action) => state + 1,
      }),
      createReducer<number, ActionType<typeof actions>>(initialState, {
        ADD: (state, action) => state + action.payload,
        INCREMENT: (state, action) => state + 1,
      }),
    ].forEach(fn => {
      // @dts-jest:pass
      fn(0, {} as any); // => 0
      // @dts-jest:pass
      fn(0, increment()); // => 1
      // @dts-jest:pass
      fn(0, add(4)); // => 4
    });
  }
}
