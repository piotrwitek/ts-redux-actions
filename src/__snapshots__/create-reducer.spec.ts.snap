// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Type refinement checks action 1`] = `"T.PayloadAction<\\"foo1\\", string>"`;

exports[`Type refinement checks action 2`] = `"T.PayloadAction<\\"foo2\\", string>"`;

exports[`Type refinement checks action 3`] = `"T.EmptyAction<\\"foo3\\">"`;

exports[`Type refinement checks action 4`] = `"T.EmptyAction<\\"foo4\\">"`;

exports[`Type refinement checks reducer.handlers 1`] = `"Record<\\"foo1\\" | \\"foo2\\" | \\"foo3\\" | \\"foo4\\", (state: { foo: string | null; }, action: T.PayloadAction<\\"foo1\\", string> | T.PayloadAction<\\"foo2\\", string> | T.EmptyAction<\\"foo3\\"> | T.EmptyAction<\\"foo4\\">) => { foo: string | null; }>"`;

exports[`Type refinement checks reducerResult 1`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks reducerResult 2`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks reducerResult 3`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks reducerResult 4`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks state 1`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks state 2`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks state 3`] = `"{ foo: string | null; }"`;

exports[`Type refinement checks state 4`] = `"{ foo: string | null; }"`;

exports[`With Action Creators counterReducer1.handlers 1`] = `"Record<\\"ADD\\" | \\"INCREMENT\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\">) => number>"`;

exports[`With Action Creators counterReducer2.handlers 1`] = `"Record<\\"ADD\\" | \\"INCREMENT\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\">) => number>"`;

exports[`With Action Creators counterReducer3.handlers 1`] = `"Record<\\"ADD\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\">) => number>"`;

exports[`With Action Creators counterReducer4.handlers 1`] = `"Record<\\"INCREMENT\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\">) => number>"`;

exports[`With Action Types counterReducer1.handlers 1`] = `"Record<\\"ADD\\" | \\"INCREMENT\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\">) => number>"`;

exports[`With Action Types counterReducer2.handlers 1`] = `"Record<\\"ADD\\" | \\"INCREMENT\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\">) => number>"`;

exports[`With Action Types counterReducer3.handlers 1`] = `"Record<\\"ADD\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\">) => number>"`;

exports[`With Action Types counterReducer4.handlers 1`] = `"Record<\\"INCREMENT\\", (state: number, action: T.PayloadAction<\\"ADD\\", number> | T.EmptyAction<\\"INCREMENT\\"> | T.EmptyAction<\\"DECREMENT\\">) => number>"`;
